lib/firmware/fsm_msg_cosmos.h

```diff
@@ -0,0 +1,182 @@
+        layoutHome();
+        return;
+    }
+
+    HDNode *node = fsm_getDerivedNode(SECP256K1_NAME, msg->address_n, msg->address_n_count, NULL);
+    if (!node) { return; }
+
+    hdnode_fill_public_key(node);
+
+    if (!confirm(ButtonRequestType_ButtonRequest_ProtectCall, _("Confirm Fee Details"), "Fee: %" PRIu32 " uATOM\nGas: %" PRIu32 "", msg->fee_amount, msg->gas))
+    {
+        fsm_sendFailure(FailureType_Failure_ActionCancelled, NULL);
+        layoutHome();
+        return;
+    }
+
+    if (!confirm(ButtonRequestType_ButtonRequest_ProtectCall, _("Confirm Aux Details"), "Memo: \"%s\"\nChain ID: %s", msg->memo, msg->chain_id))
+    {
+        fsm_sendFailure(FailureType_Failure_ActionCancelled, NULL);
+        layoutHome();
+        return;
+    }
+
+    RESP_INIT(CosmosMsgRequest);
+
+    if (!cosmos_signTxInit(node,
+                           msg->address_n,
+                           msg->address_n_count,
+                           msg->account_number,
+                           msg->chain_id,
+                           strlen(msg->chain_id),
+                           msg->fee_amount,
+                           msg->gas,
+                           msg->has_memo ? msg->memo : "",
+                           msg->has_memo ? strlen(msg->memo) : 0,
+                           msg->sequence,
+                           msg->msg_count))
+    {
+        fsm_sendFailure(FailureType_Failure_FirmwareError,
+                        _("Failed to initialize transaction signing"));
+        layoutHome();
+        return;
+    }
+
+
+
+    layoutHome();
+    msg_write(MessageType_MessageType_CosmosMsgRequest, resp);
+}
+
+void fsm_msgCosmosMsgAck(const CosmosMsgAck* msg) {
+    // Confirm transaction basics
+    CHECK_PARAM(cosmos_signingIsInited(), "Must call CosmosSignTx to initiate signing");
+    if (!msg->has_send || !msg->send.has_from_address || !msg->send.has_to_address || !msg->send.has_amount) {
+        cosmos_signAbort();
+        fsm_sendFailure(FailureType_Failure_FirmwareError,
+                        _("Invalid Cosmos Message Type"));
+        layoutHome();
+        return;
+    }
+
+    const char *coin_name = "Cosmos";
+    const CoinType *coin = fsm_getCoin(true, coin_name);
+    if (!coin) { return; }
+
+    uint32_t address_n[8];
+    if(!cosmos_getAddressN(address_n, 8)) {
+        cosmos_signAbort();
+        fsm_sendFailure(FailureType_Failure_FirmwareError, "Failed to get derivation path");
+        return;
+    }
+    size_t address_n_count = cosmos_getAddressNCount();
+    char node_str[NODE_STRING_LENGTH];
+    if (!bip32_node_to_string(node_str, sizeof(node_str), coin, address_n,
+                              address_n_count, /*whole_account=*/false,
+                              /*show_addridx=*/true) &&
+        !bip32_path_to_string(node_str, sizeof(node_str),
+                              address_n, address_n_count)) {
+        memset(node_str, 0, sizeof(node_str));
```

Comment:
I think we need to `fsm_sendFailure` after the memset, since otherwise there isn't a human-readable account description in `node_str`.